Framework/src/BookkeepingQualitySink.cxx

```diff
@@ -0,0 +1,139 @@
+        .to = flag.getEnd() == gFullValidityInterval.getMax() ? std::nullopt : std::optional<uint64_t>{ flag.getEnd() },
+        .origin = flag.getSource(),
+        .comment = flag.getComment() });
+    }
+
+    if (bkpQcFlags.empty()) {
+      continue;
+    }
+
+    try {
+      switch (type) {
+        case Provenance::SyncQC:
+        case Provenance::AsyncQC:
+          qcClient->createForDataPass(flagCollection->getRunNumber(), flagCollection->getPassName(), flagCollection->getDetector(), bkpQcFlags);
+          break;
+        case Provenance::MCQC:
+          qcClient->createForSimulationPass(flagCollection->getRunNumber(), flagCollection->getPeriodName(), flagCollection->getDetector(), bkpQcFlags);
+          break;
+      }
+    } catch (const std::runtime_error& err) {
+      ILOG(Error, Support) << "Failed to send flags for detector: " << flagCollection->getDetector()
+                           << " and pass: " << flagCollection->getPassName()
+                           << " with error: " << err.what() << ENDM;
+    }
+  }
+}
+
+BookkeepingQualitySink::BookkeepingQualitySink(const std::string& grpcUri, Provenance type, SendCallback sendCallback)
+  : mGrpcUri{ grpcUri }, mType{ type }, mSendCallback{ sendCallback } {}
+
+auto merge(std::unique_ptr<QualityControlFlagCollection>&& collection, const std::unique_ptr<QualityObject>& qualityObject) -> std::unique_ptr<QualityControlFlagCollection>
+{
+  QualitiesToFlagCollectionConverter converter(std::move(collection), qualityObject->getPath());
+  converter(*qualityObject);
+  return converter.getResult();
+}
+
+auto collectionFromQualityObject(const QualityObject& qualityObject) -> std::unique_ptr<QualityControlFlagCollection>
+{
+  return std::make_unique<QualityControlFlagCollection>(
+    qualityObject.getName(),
+    qualityObject.getDetectorName(),
+    gFullValidityInterval,
+    qualityObject.getActivity().mId,
+    qualityObject.getActivity().mPeriodName,
+    qualityObject.getActivity().mPassName,
+    qualityObject.getActivity().mProvenance);
+}
+
+auto to_key(const QualityObject& qualityObject) -> std::string
+{
+  std::string key;
+  key.reserve(qualityObject.getDetectorName().size() + std::size("||") + qualityObject.getCheckName().size());
+  key.append(qualityObject.getDetectorName());
+  key.append("||");
+  key.append(qualityObject.getCheckName());
+  return key;
+}
+
+template <typename t>
+struct tt;
+
+void BookkeepingQualitySink::run(framework::ProcessingContext& context)
+{
+  for (auto const& ref : framework::InputRecordWalker(context.inputs())) {
+    try {
+      auto qualityObject = framework::DataRefUtils::as<QualityObject>(ref);
+      auto [emplacedIt, _] = mQualityObjectsMap.emplace(to_key(*qualityObject), collectionFromQualityObject(*qualityObject));
+      emplacedIt->second = merge(std::move(emplacedIt->second), qualityObject);
```

Comment:
Please store and send to BKP one FlagCollection per detector, not one per QO. Indeed it must have been lost in the requirements I passed to you - we agreed some time ago with Martin to have one upload call per detector, to minimize the overhead on BKP.

Reply:
I don't really understand how to do that. The BKP API has passname/productionname and detector as parameters. So I cannot send everything per one detector. only I need to separate by these two parameters. Which is also the reason why I am using the `unordered_map`: to keep flag collections separated by their names and detectors. So this is already an optimization to reduce the number of objects sent into the BKP.

Reply:
Solved in person. The issue was my misunderstanding of the nomenclature. I will rewrite the code, so everything is grouped by detector only.